### PR TITLE
Optimizer

### PR DESCRIPTION
**Pull request by guilhermebodin**
Hey guys I just created this PR to get some help in implementing the Optimizer, for now not all tests are passing for `contlinear` and `contconic`.

### CODE REVIEW

File: src/optimizer.jl [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                              MOI.ObjectiveFunction{MOI.SingleVariable},
+                              MOI.ObjectiveFunction{MOI.ScalarAffineFunction{Float64}}})
+    return true
+end
+
+function MOI.supports_constraint(optimizer::DualOptimizer, F::Type{<:SF}, S::Type{<:SS})
+    if optimizer.dual_optimizer !== nothing
+        MOI.supports_constraint(optimizer.dual_optimizer, F, S)
+    else
+        return true
```
Comment (blegat) [inline src/optimizer.jl]: This is weird, maybe you should not allow the `dual_optimizer` and `dual_problem` to be `nothing`

File: src/optimizer.jl [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+const SS = Union{MOI.EqualTo{Float64}, MOI.GreaterThan{Float64}, MOI.LessThan{Float64}, 
+                 MOI.Zeros, MOI.Nonnegatives, MOI.Nonpositives, 
+                 MOI.SecondOrderCone, MOI.RotatedSecondOrderCone,
+                 MOI.ExponentialCone, MOI.DualExponentialCone,
+                 MOI.PowerCone, MOI.DualPowerCone,
+                 MOI.PositiveSemidefiniteConeTriangle}
+
+mutable struct DualOptimizer <: MOI.AbstractOptimizer
+    dual_problem::Union{Nothing, Dualization.DualProblem}
+    dual_optimizer::Union{Nothing, MOI.AbstractOptimizer}
```
Comment (blegat) [inline src/optimizer.jl]: I would do `dual_optimizer::OT` et `DualOptimizer{OT}`.
Otherwise, accessing this field is type instable and it will be slow

Comment (joaquimg) [reply src/optimizer.jl]: yup, that makes sense!

File: src/optimizer.jl (hunk 1 above)
Comment (joaquimg) [reply src/optimizer.jl]: so, for now the dual model is being built to `dual_problem` and its only passes to `dual_optimizer` when `optimize!` is called.
One possible next step is get rid of the `dual_optimizer` completely and plug the optimizer inside the `dual_problem`, in the `dual_model` field.

File: src/optimizer.jl (hunk 2 above)
Comment (guilhermebodin) [reply src/optimizer.jl]: what does OT stands for?

Comment (blegat) [reply src/optimizer.jl]: **O**ptimizer **T**ype

File: src/optimizer.jl [hunk 3]
```diff
@@ -107,32 +107,68 @@ end
...
+    for (i, vi) in enumerate(vis)
+        dual_optimizer_vi[i] = varmap[vi]
+    end
+    return dual_optimizer_vi
 end
 
 function MOI.get(optimizer::DualOptimizer, ::MOI.ConstraintPrimal, 
                  ci::CI{F,S}) where {F <: MOI.AbstractScalarFunction, S}
     ci_dual_problem = optimizer.dual_problem.primal_dual_map.primal_con_dual_con[ci]
+<<<<<<< HEAD
```
Comment (blegat) [inline src/optimizer.jl]: merge coflict

File: src/optimizer.jl (hunk 2 above)
Comment (guilhermebodin) [reply src/optimizer.jl]: I tried like this but it doesn't work
```
mutable struct DualOptimizer{OT <: MOI.ModelLike} <: MOI.AbstractOptimizer
    dual_problem::Union{Nothing, DualProblem}
    dual_optimizer::OT
    dual_optimizer_idx_map::Union{Nothing, MOIU.IndexMap}

    function DualOptimizer{OT}(dual_optimizer::OT) where {OT <: MOI.ModelLike}
        return new(nothing, dual_optimizer, nothing)
    end
end
```
It gives me and error
`ERROR: LoadError: LoadError: MethodError: no method matching Dualization.DualOptimizer(::GLPK.Optimizer)`

Comment (joaquimg) [reply src/optimizer.jl]: I think it should be:
```julia
    function DualOptimizer{OT}(dual_optimizer::OT) where {OT <: MOI.ModelLike}
        return new(nothing, dual_optimizer, nothing)
    end
    function DualOptimizer(dual_optimizer::OT) where {OT <: MOI.ModelLike}
        return DualOptimizer{OT}(dual_optimizer)
    end
```

Comment (guilhermebodin) [reply src/optimizer.jl]: true, just saw what I missed, the constructor should not have '{OT}'

**Comment by guilhermebodin** [comment]
This closes #8 #18 #16

**Pull request merged**